src/libsyntax/parse/literal.rs

```diff
@@ -0,0 +1,487 @@
+                LitKind::ByteStr(Lrc::new(buf))
+            }
+            token::ByteStrRaw(i, _) => {
+                LitKind::ByteStr(Lrc::new(i.to_string().into_bytes()))
+            }
+        })
+    }
+
+    /// Attempts to recover a token from semantic literal.
+    /// This function is used when the original token doesn't exist (e.g. the literal is created
+    /// by an AST-based macro) or unavailable (e.g. from HIR pretty-printing).
+    pub fn to_lit_token(&self) -> (token::Lit, Option<Symbol>) {
+        match *self {
+            LitKind::Str(string, ast::StrStyle::Cooked) => {
+                let escaped = string.as_str().escape_default().to_string();
+                (token::Lit::Str_(Symbol::intern(&escaped)), None)
+            }
+            LitKind::Str(string, ast::StrStyle::Raw(n)) => {
+                (token::Lit::StrRaw(string, n), None)
+            }
+            LitKind::ByteStr(ref bytes) => {
+                let string = bytes.iter().cloned().flat_map(ascii::escape_default)
+                    .map(Into::<char>::into).collect::<String>();
+                (token::Lit::ByteStr(Symbol::intern(&string)), None)
+            }
+            LitKind::Byte(byte) => {
+                let string: String = ascii::escape_default(byte).map(Into::<char>::into).collect();
+                (token::Lit::Byte(Symbol::intern(&string)), None)
+            }
+            LitKind::Char(ch) => {
+                let string: String = ch.escape_default().map(Into::<char>::into).collect();
+                (token::Lit::Char(Symbol::intern(&string)), None)
+            }
+            LitKind::Int(n, ty) => {
+                let suffix = match ty {
+                    ast::LitIntType::Unsigned(ty) => Some(Symbol::intern(ty.ty_to_string())),
+                    ast::LitIntType::Signed(ty) => Some(Symbol::intern(ty.ty_to_string())),
+                    ast::LitIntType::Unsuffixed => None,
+                };
+                (token::Lit::Integer(Symbol::intern(&n.to_string())), suffix)
+            }
+            LitKind::Float(symbol, ty) => {
+                (token::Lit::Float(symbol), Some(Symbol::intern(ty.ty_to_string())))
+            }
+            LitKind::FloatUnsuffixed(symbol) => (token::Lit::Float(symbol), None),
+            LitKind::Bool(value) => {
+                let kw = if value { keywords::True } else { keywords::False };
+                (token::Lit::Bool(kw.name()), None)
+            }
+            LitKind::Err(val) => (token::Lit::Err(val), None),
+        }
+    }
+}
+
+impl Lit {
+    /// Converts literal token with a suffix into an AST literal.
+    /// Works speculatively and may return `None` is diagnostic handler is not passed.
+    /// If diagnostic handler is passed, may return `Some`,
+    /// possibly after reporting non-fatal errors and recovery, or `None` for irrecoverable errors.
+    crate fn from_token(
+        token: &token::Token,
+        span: Span,
+        diag: Option<(Span, &Handler)>,
```

Comment:
`diag: Option` API feels awkward. From my reading of the code, the following is true

* `diag` is used **only** to report invalid suffix errors
* the fist call site is attrs/mod, where diag is always `None`
* the second call site is in this module, where diag is always `Some`

Perhaps we could remove `diag` argument altogether, and instead add an `fn validate_suffix(literal, suffix, span, handler)` function which we call from the parser? 

On the higher level, perhaps the problem is that attrs/mod doesn't have a diagnostic handler? It's not obvious why is it ok to ignore diagnostics there...

Reply:
>On the higher level, perhaps the problem is that attrs/mod doesn't have a diagnostic handler?

The historical reason, AFAIK, is that meta-items can be parsed during command line parsing or pretty-printing and `Parser` (and `ParseSess` maybe) are unavailable at that time.
Perhaps things have changed because I see that parser/parse-session are created e.g. in `fn parse_cfgspecs` now.

If it's possible to thread parse session through all the meta-item parsing code that would be great, because `attrs/mod.rs` contains a lot of code that duplicates normal parsing.
That's a pretty large refactoring though, `my_attr.meta()` is used during processing of pretty much every built-in attribute and it doesn't use parse session.

>It's not obvious why is it ok to ignore diagnostics there...

Meta-item parsing works speculatively and if the result is `None` it reports one big "ill-formed meta item" error.

Reply:
>diag is used only to report invalid suffix errors

There's also the "int literal is too large" error for literals not fitting into `u128`.

>Perhaps we could remove diag argument altogether, and instead add an fn validate_suffix(literal, suffix, span, handler) function which we call from the parser?

The suffix checking is somewhat intertwined with lowering `(token::Lit, Suffix) -> ast::LitKind` because lowering can turn integer literal token with suffix like `1f32` into floating point AST literals (see `fn integer_lit/float_lit/filtered_float_lit`), so I just kept those functions alone and didn't change them.
Right now we need the lowering part (as opposed to the validation part) anyway, because AST still contains `LitKind`, perhaps we can reconsider and split lowering and validation later after lowering only happens during HIR construction and after attr/mod has access to `Handler` (if possible).

Reply:
Yeah, makes sense to postpone this. Another alternative would be to change from `diag` input to `Result<LitKind, Error>` and feed `Error` into `Handler` in the parser and ignore in meta items, like unescaping works, but I imagine this won't be really better, given that the code in question is in the single module anyway